pkg/tunneler/tunnel.go

```diff
@@ -105,138 +109,120 @@ func SyncerTunnelURL(host, ws, target string) (string, error) {
 	if err != nil || hostURL.Scheme != "https" || hostURL.Host == "" {
 		return "", fmt.Errorf("wrong url format, expected https://host<:port>/<path>: %w", err)
 	}
-	host = strings.Trim(host, "/")
-	return host + defaultTunnelPathPrefix + "/" + ws + "/apis/" + workloadv1alpha1.SchemeGroupVersion.String() + "/synctargets/" + target, nil
+	return url.JoinPath(hostURL.String(), "clusters", ws, "apis", workloadv1alpha1.SchemeGroupVersion.String(), "synctargets", target)
 }
 
-// WithSyncerTunnel adds an HTTP Handler that handles reverse connections and reverse proxy requests using 2 different paths:
+// WithSyncerTunnelHandler adds an HTTP Handler that handles reverse connections via the tunnel subresource:
 //
-// https://host/services/syncer-tunnels/clusters/<ws>/apis/workload.kcp.io/v1alpha1/synctargets/<name>/connect establish reverse connections and queue them so it can be consumed by the dialer
-// https://host/services/syncer-tunnels/clusters/<ws>/apis/workload.kcp.io/v1alpha1/synctargets/<name>/proxy/{path} proxies the {path} through the reverse connection identified by the cluster and syncer name
-func WithSyncerTunnel(apiHandler http.Handler) http.HandlerFunc {
-	pool := newTunnelPool()
+// https://host/clusters/<ws>/apis/workload.kcp.io/v1alpha1/synctargets/<name>/tunnel establish reverse connections and queue them so it can be consumed by the dialer
+func (tn *Tunneler) WithSyncerTunnelHandler(apiHandler http.Handler) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		// fall through, syncer tunnels URL start by /services/tunnels
-		if !strings.HasPrefix(r.URL.Path, defaultTunnelPathPrefix) {
+		logger := klog.FromContext(r.Context())
+
+		ri, ok := genericapirequest.RequestInfoFrom(r.Context())
+		if !ok {
 			apiHandler.ServeHTTP(w, r)
 			return
 		}
 
-		// route the request
-		p := strings.TrimPrefix(r.URL.Path, defaultTunnelPathPrefix)
-		path := strings.Split(strings.Trim(p, "/"), "/")
-		if len(path) < 7 {
-			http.Error(w, "invalid path", http.StatusInternalServerError)
+		if !ri.IsResourceRequest ||
+			ri.Resource != "synctargets" ||
+			ri.Subresource != "tunnel" ||
+			ri.APIGroup != workloadv1alpha1.SchemeGroupVersion.Group ||
+			ri.APIVersion != workloadv1alpha1.SchemeGroupVersion.Version {
+			apiHandler.ServeHTTP(w, r)
 			return
 		}
 
-		gv := workloadv1alpha1.SchemeGroupVersion
-		if path[1] != "apis" ||
-			path[2] != gv.Group ||
-			path[3] != gv.Version ||
-			path[4] != "synctargets" {
+		// Split the RequestURI to extract clusterName and SyncerName from the full path.
+		path := strings.Split(r.RequestURI, "/")
```

Comment:
Not sure this Splitting is necessary. Don't you already have the information in the `RequestInfo.Parts` field (or possibly `RequestInfo.Path`  ?

Reply:
There is also the `RequestInfo.Name`

Reply:
Done!